### ISSUE TITLE
equalize_clahe slow path is not transposition-equivariant below float64 when clipping

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4632]
On the differentiable path with a clip limit, `equalize_clahe` is not equivariant under transposing the
image and the grid — even for a **square** grid, where no axis asymmetry is involved. The output moves by
up to `2.9e-2` in `bfloat16`.

## Reproduction

```python
import torch
from kornia import enhance

for dt in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    bad, worst = 0, 0.0
    for s in range(50):
        torch.manual_seed(s)
        img = torch.rand(1, 1, 12, 12, dtype=dt)
        a = enhance.equalize_clahe(img, 40.0, (2, 2), True)
        b = enhance.equalize_clahe(img.transpose(-2, -1), 40.0, (2, 2), True).transpose(-2, -1)
        d = (a.float() - b.float()).abs().max().item()
        worst = max(worst, d)
        bad += d > 1e-3
    print(dt, bad, worst)
```

Measured on `main` (`3f748920e`), torch 2.14.0, CPU, macOS arm64:

| dtype | seeds past `1e-3` | worst abs diff |
| --- | --- | --- |
| float64 | 0/50 | `0.000e+00` |
| float32 | 0/50 | `1.192e-07` |
| **bfloat16** | **48/50** | **`2.930e-02`** |
| float16 | 0/50 | `9.766e-04` |

float64 is exact and float32 is at the rounding floor, so this is precision-dependent rather than an
indexing bug. bfloat16 is affected on nearly every seed; float16 sits just under the threshold but its
worst case is already `9.8e-4`, i.e. the same phenomenon a factor of 30 smaller.

## Likely cause

The clip-and-redistribute step sums a soft histogram and then treats the excess as a count. Below
float64 a rounding residue in that sum can cross an integer boundary, which changes how much mass is
redistributed — and the residue depends on summation order, which transposition changes. The fast path
does not show it because it histograms with `bincount` on integers.

Not a regression: identical on `main` and on #4628's merge tree. Found while reviewing #4628, which
adds a transposition test — that test deliberately runs the slow path at `clip_limit=0` to avoid this,
and the reason should be this issue rather than a comment.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by NiteeshGowda72** [comment]
I am working on it.

**Comment by awss1i** [comment]
Adding the Linux x86_64 side, since the scope here depends on platform and input size. I used the same loop as the reproduction above: 50 seeds, `clip_limit=40`, `slow_and_differentiable=True`, counting seeds past `1e-3`. Run on `main` at `5be74dc9`, torch 2.14.0+cpu, Python 3.11, Linux x86_64:

| input, grid | float64 | float32 | float16 | bfloat16 |
| --- | --- | --- | --- | --- |
| `1x1x12x12`, `(2, 2)` | 0/50 | 2/50, worst `2.778e-02` | 1/50, worst `2.832e-02` | 49/50, worst `3.906e-03` |
| `1x1x24x24`, `(2, 2)` | 0/50 | 13/50, worst `6.945e-03` | 10/50, worst `7.812e-03` | 50/50, worst `1.172e-02` |
| `1x1x60x60`, `(4, 4)` | 0/50 | 46/50, worst `3.906e-03` | 43/50, worst `4.395e-03` | 50/50, worst `7.812e-03` |
| `2x3x60x60`, `(4, 4)` | 0/50 | 50/50, worst `3.906e-03` | 50/50, worst `4.395e-03` | 50/50, worst `9.766e-03` |
| `1x3x64x64`, `(8, 8)` | 0/50 | 49/50, worst `1.563e-02` | 49/50, worst `1.611e-02` | 50/50, worst `1.953e-02` |

So float32 and float16 are affected as well, not only bfloat16. On the issue's own `1x1x12x12` input they cross `1e-3` on 2 and 1 of 50 seeds here, and on larger inputs on almost every seed. Only float64 is exact.

On the cause, I traced it while writing the transposition test in #4628 (float32, grid `(1, 2)` on `1x2x12x24`, clip 40). In every failing seed, `clipped = pixels - histos.sum(1)` differed between the two pixel orders, in one of two ways:
- **28 of 44 seeds:** exactly `0` in one order and a residue such as `3.052e-05` in the other, so `mat_range < residual` added a count to the first bin.
- **16 of 44 seeds:** the residue was negative in one order, and `torch.remainder(clipped, 256)` wrapped it to about 256, adding a count to every bin.

Either way the lookup table moves, so the output does too.
